python/pyspark/mllib/linalg/__init__.py

```diff
@@ -1296,9 +1296,19 @@ def asML(self):
         return newlinalg.SparseMatrix(self.numRows, self.numCols, self.colPtrs, self.rowIndices,
                                       self.values, self.isTransposed)
 
-    # TODO: More efficient implementation:
     def __eq__(self, other):
-        return np.all(self.toArray() == other.toArray())
+        if ((type(other) is not type(self)) or
```

Comment:
So this will return false for a spare matrix and a dense matrix being equal to eachother even if all of the elements are equal, and previously this was not the case. This also doesn't match the Scala implementation. Ping @dbtsai who has been doing some of the mllib-local/linalg stuff :)